feature/dsql/token/auth_token_generator.go

```diff
@@ -0,0 +1,121 @@
+// * region - Region is where the database is located
+// * creds - Credentials to use when signing the token
+func GenerateDbConnectAuthToken(ctx context.Context, endpoint, region string, creds aws.CredentialsProvider, optFns ...func(options *AuthTokenOptions)) (string, error) {
+	values := url.Values{
+		"Action": []string{userAction},
+	}
+	return generateAuthToken(ctx, endpoint, region, values, rdsClusterTokenID, creds, optFns...)
+}
+
+// GenerateDBConnectAdminAuthToken Generates an admin authentication token for IAM authentication to a DSQL database.
+//
+// This is the admin user variant, see [GenerateDbConnectAuthToken] for the regular user variant
+//
+// * endpoint - Endpoint is the hostname and optional port to connect to the database
+// * region - Region is where the database is located
+// * creds - Credentials to use when signing the token
+func GenerateDBConnectAdminAuthToken(ctx context.Context, endpoint, region string, creds aws.CredentialsProvider, optFns ...func(options *AuthTokenOptions)) (string, error) {
+	values := url.Values{
+		"Action": []string{adminUserAction},
+	}
+	return generateAuthToken(ctx, endpoint, region, values, rdsClusterTokenID, creds, optFns...)
+}
+
+// All generate token functions are presigned URLs behind the scenes with the scheme stripped.
+// This function abstracts generating this for all use cases
+func generateAuthToken(ctx context.Context, endpoint, region string, values url.Values, signingID string, creds aws.CredentialsProvider, optFns ...func(options *AuthTokenOptions)) (string, error) {
+	if len(region) == 0 {
+		return "", fmt.Errorf("region is required")
+	}
+	if len(endpoint) == 0 {
+		return "", fmt.Errorf("endpoint is required")
+	}
+
+	o := AuthTokenOptions{}
+
+	for _, fn := range optFns {
+		fn(&o)
+	}
+
+	if o.ExpiresIn == 0 {
+		o.ExpiresIn = 15 * time.Minute
+	}
+
+	if creds == nil {
+		return "", fmt.Errorf("credetials provider must not ne nil")
+	}
+
+	// the scheme is arbitrary and is only needed because validation of the URL requires one.
+	if !(strings.HasPrefix(endpoint, "http://") || strings.HasPrefix(endpoint, "https://")) {
+		endpoint = "https://" + endpoint
+	}
+
+	req, err := http.NewRequest("GET", endpoint, nil)
+	if err != nil {
+		return "", err
+	}
+	req.URL.RawQuery = values.Encode()
+	signer := v4.NewSigner()
+
+	credentials, err := creds.Retrieve(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	expires := o.ExpiresIn
+	// if credentials expire before expiresIn, set that as the expiration time
+	if credentials.CanExpire && !credentials.Expires.IsZero() {
+		credsExpireIn := credentials.Expires.Sub(sdk.NowTime())
+		expires = min(o.ExpiresIn, credsExpireIn)
+	}
+	query := req.URL.Query()
+	query.Set("X-Amz-Expires", strconv.Itoa(int(expires.Seconds())))
+	req.URL.RawQuery = query.Encode()
+
+	signedURI, _, err := signer.PresignHTTP(ctx, credentials, req, emptyPayloadHash, signingID, region, sdk.NowTime().UTC())
```

Comment:
do we know why empty payload hash is needed? other SDKs did not need this.

Reply:
They probably default to an empty payload hash internally if none is set. The signer on Go v2 mentions explicitly that an empty hash needs to be passed along.